dissect/protos/dns.py

```diff
@@ -0,0 +1,575 @@
+    DNS_TYPE_KX    : 'KX',
+    DNS_TYPE_CERT  : 'CERT',
+    DNS_TYPE_DNAME : 'DNAME',
+    DNS_TYPE_OPT   : 'OPT',
+    DNS_TYPE_APL   : 'APL',
+    DNS_TYPE_DS    : 'DS',
+    DNS_TYPE_SSHFP : 'SSHFP',
+    DNS_TYPE_IPSECKEY : 'IPSECKEY',
+    DNS_TYPE_RRSIG : 'RRSIG',
+    DNS_TYPE_NSEC  : 'NSEC',
+    DNS_TYPE_DNSKEY: 'DNSKEY',
+    DNS_TYPE_DHCID : 'DHCID',
+    DNS_TYPE_NSEC3 : 'NSEC3',
+    DNS_TYPE_NSEC3PARAM : 'NSEC3PARAM',
+    DNS_TYPE_TLSA  : 'TLSA',
+    DNS_TYPE_HIP   : 'HIP',
+    DNS_TYPE_CDS   : 'CDS',
+    DNS_TYPE_CDNSKEY : 'CDNSKEY',
+    DNS_TYPE_TKEY  : 'TKEY',
+    DNS_TYPE_TSIG  : 'TSIG',
+    DNS_TYPE_IXFR  : 'IXFR',
+    DNS_TYPE_AXFR  : 'AXFR',
+    DNS_TYPE_ANY   : 'ANY',
+    DNS_TYPE_URI   : 'URI',
+    DNS_TYPE_CAA   : 'CAA',
+    DNS_TYPE_TA    : 'TA',
+    DNS_TYPE_DLV   : 'DLV',
+}
+
+DNS_CLASS_IN     = 1
+DNS_CLASS_CSNET  = 2
+DNS_CLASS_CHAOS  = 3
+DNS_CLASS_HESIOD = 4
+DNS_CLASS_ANY    = 255
+
+dns_class_names = {
+    DNS_CLASS_IN     : 'IN',
+    DNS_CLASS_CSNET  : 'CSNET',
+    DNS_CLASS_CHAOS  : 'CHAOS',
+    DNS_CLASS_HESIOD : 'HESIOD',
+    DNS_CLASS_ANY    : 'ANY',
+}
+
+DNS_NAMETYPE_LABEL        = 0
+DNS_NAMETYPE_RESERVED     = 1
+DNS_NAMETYPE_EXTENDED     = 2 
+DNS_NAMETYPE_POINTER      = 3
+DNS_NAMETYPE_LABELPOINTER = 4
+
+DNS_FLAG_RESP   = 1 << 15
+DNS_FLAG_TRUNC  = 1 << 9
+DNS_FLAG_RECUR  = 1 << 8
+DNS_FLAG_AD     = 1 << 5
+#DNS_FLAG_AA = 1 << 5    # authoritative answer
+#DNS_FLAG_TC = 1 << 6    # truncated response
+#DNS_FLAG_RD = 1 << 7    # recursion desired
+#DNS_FLAG_RA = 1 << 8    # recursion allowed
+#DNS_FLAG_CD = 1 << 11   # checking disabled
+
+#totally arbitrary count value to abort parsing dns records
+DNS_SUSPICIOUS_COUNT = 0x20
+
+class DnsParseError(Exception):
+    pass
+
+def dnsFlagsOp(val):
+    return (val >> 11) & 0xf
+
+class DnsNameLabel(VStruct):
+    '''
+    A DNS Name component.
+    '''
+    def __init__(self, labelbytes=''):
+        VStruct.__init__(self)
+        labelbytes =str(labelbytes).encode('utf-8')
+        self._nametype = 0
+        self._pointer  = 0
+        self.length = uint8( endian='big' ).vsOnset(self.pcb_length)
+        self.length = len(labelbytes)
+        self.label  = vbytes(len(labelbytes),labelbytes).vsOnset(self.pcb_label)
+
+    def pcb_length(self):
+        size = self.length
+        labeltype = size >> 6 # if it's over 63 then it should be a pointer
+        # ordered by use
+        if labeltype == 0b11:
+            self._nametype = DNS_NAMETYPE_POINTER
+            size = 1
+        elif labeltype == 0b00:
+            self._nametype = DNS_NAMETYPE_LABEL
+        elif labeltype == 0b10:
+            raise DnsParseError('Extended labeltype is not supported.') 
+        elif labeltype == 0b01:
+            raise DnsParseError('Unrecognized labeltype (reserved).')
+        if self.vsHasField('label'):  # the resize will crash if the label hasn't been set yet w/o this
+            self.vsGetField('label').vsResize(size)
+
+
```

Comment:
extra empty line ( which totally may have been inherited from my previous DNS cruft, but lets kill it off since we're touching it... )